components/script/dom/windowproxy.rs

```diff
@@ -200,6 +204,134 @@ impl WindowProxy {
+                opener_pipeline_id: self.currently_active.get().unwrap(),
+                new_browsing_context_id: new_browsing_context_id,
+                new_top_level_browsing_context_id: new_top_level_browsing_context_id,
+                new_pipeline_id: new_pipeline_id,
+            };
+            let document = self.currently_active.get()
+                .and_then(|id| ScriptThread::find_document(id))
+                .unwrap();
+            let blank_url = ServoUrl::parse("about:blank").ok().unwrap();
+            let load_data = LoadData::new(blank_url,
+                                          None,
+                                          document.get_referrer_policy(),
+                                          Some(document.url().clone()));
+            let (pipeline_sender, pipeline_receiver) = ipc::channel().unwrap();
+            let new_layout_info = NewLayoutInfo {
+                parent_info: None,
+                new_pipeline_id: new_pipeline_id,
+                browsing_context_id: new_browsing_context_id,
+                top_level_browsing_context_id: new_top_level_browsing_context_id,
+                load_data: load_data,
+                pipeline_port: pipeline_receiver,
+                content_process_shutdown_chan: None,
+                window_size: None,
+                layout_threads: PREFS.get("layout.threads").as_u64().expect("count") as usize,
+            };
+            let constellation_msg = ScriptMsg::ScriptNewAuxiliary(load_info, pipeline_sender);
+            window.send_to_constellation(constellation_msg);
+            ScriptThread::process_attach_layout(new_layout_info, document.origin().clone());
+            let msg = EmbedderMsg::BrowserCreated(new_top_level_browsing_context_id);
+            window.send_to_embedder(msg);
+            let auxiliary = ScriptThread::find_document(new_pipeline_id).and_then(|doc| doc.browsing_context());
+            if let Some(proxy) = auxiliary {
+                if name.to_lowercase() != "_blank" {
+                    proxy.set_name(name);
+                }
+                return Some(proxy)
+            }
+        }
+        None
+    }
+
+    // https://html.spec.whatwg.org/multipage/#window-open-steps
+    pub fn open(&self,
+                url: DOMString,
+                target: DOMString,
+                features: DOMString)
+                -> Option<DomRoot<WindowProxy>> {
+        // Step 3.
+        let non_empty_target = match target.as_ref() {
+            "" => DOMString::from("_blank"),
+            _ => target
+        };
+        // TODO Step 4, properly tokenize features.
+        let noopener = features.contains("noopener");
+        let (chosen, new)  = match self.choose_browsing_context(non_empty_target, noopener) {
+            (Some(window_proxy), new) => (window_proxy, new),
+            (None, _) => return None
+        };
+        if let Some(target_document) = chosen.document() {
```

Comment:
It would be more readable to do:

```rust
let document = match chosen.document() { Some(doc) => doc, None => return };
```

That would let you early return and not have to `return Some(_)` at the end of the block. The code flow of this method is difficult to follow.

Reply:
good one, thanks